### PR TITLE
[builtin/history] Add support for '-a' (append) flag

### PR DESCRIPTION
**Pull request by PossiblyAShrub**
Now we can do something like this:

```sh
$ echo 'start of osh session'
start of osh session
$ cat ~/.config/oil/history_osh
$ history -a
$ cat ~/.config/oil/history_osh
echo 'start of osh session'
cat ~/.config/oil/history_osh
history -a
```

### CODE REVIEW

**Comment by andychu** [comment]
This looks great!  I'll demo adding a spec test during our meeting tomorrow

https://github.com/oilshell/oil/wiki/Spec-Tests

**Comment by andychu** [comment]
BTW the `test/spec-bin.sh` installation might help with the bash issue

https://github.com/oilshell/oil/wiki/Spec-Tests

It makes sure we have a specific version of bash to test

on my laptop, I think we were falling back to the system bash and getting weird results, but our own compiled version is probably more predictable

**Comment by PossiblyAShrub** [comment]
This should be ready for review. I am not sure why the ovm-tarball test is failing. The logs say "all tests passed". Any ideas?

**Comment by andychu** [comment]
Hm I merged master and it passed

I thought that only soil-staging was broken last night, not master, not sure why this branch got broken ?

Anyway I will look at this now

File: core/shell.py [hunk 1]
```diff
@@ -378,6 +380,15 @@ def Main(lang, arg_r, environ, login_shell, loader, readline):
...
   dir_stack = state.DirStack()
 
+  # Find common directories, these are consumed by some builtins and then much
+  # later used while setting up an interactive shell.
+  home_dir = pyos.GetMyHomeDir()
+  assert home_dir is not None
+
+  # init the HISTFILE variable to our default history file
+  history_filename = os_path.join(home_dir, '.config/oil/history_%s' % lang)
+  mem.SetValue(location.LName('HISTFILE'), value.Str(history_filename), scope_e.GlobalOnly)
```
Comment (andychu) [inline core/shell.py:390]: There's a helper `state.SetGlobalString('HISTFILE', history_filename)` you can use here

File: spec/builtins2.test.sh [hunk 2]
```diff
@@ -121,6 +121,101 @@ command command -v seq
 ## N-I zsh stdout-json: ""
 ## N-I zsh status: 127
 
+#### history -a
+case $SH in dash|mksh|zsh) exit 0 ;; esac
```
Comment (andychu) [inline spec/builtins2.test.sh:125]: After this PR we should probably split this into  `spec/builtins-history.test.sh`

If a feature is only implemented in bash, and not other shells, then that's a good reason to have another file

I just did that with `spec/builtin-trap.test.sh` and `spec/builtin-trap-bash.test.sh`

File: spec/builtins2.test.sh [hunk 3]
```diff
@@ -121,6 +121,101 @@ command command -v seq
...
+echo 1
+history -a
+cat tmp
+
+echo 2
+
+cat tmp
+' | $SH -i
+
+# match osh's behaviour of echoing ^D for EDF
```
Comment (andychu) [inline spec/builtins2.test.sh]: typo EDF -> EOF

**Review by andychu — COMMENTED** [review]
Thanks, this looks great!

Just one small improvement

**Comment by PossiblyAShrub** [comment]
All done resolving your review feedback.

**Pull request merged**